### PR TITLE
Use NNPACK for strided convolutions.

### PR DESCRIPTION
**Pull request by AshkanAliabadi**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27402 Use NNPACK for strided convolutions.**

Differential Revision: [D18126265](https://our.internmc.facebook.com/intern/diff/D18126265)

### CODE REVIEW

File: aten/src/ATen/native/Convolution.cpp [hunk 1]
```diff
@@ -648,7 +647,7 @@ at::Tensor _convolution(
   } else if (input.device().type() == c10::DeviceType::CPU || input.device().type() == c10::DeviceType::CUDA) {
     if (params.use_cpu_depthwise3x3_winograd(input, weight)) {
       output = convolution_depthwise3x3_winograd_stub(
-        input.device().type(), input, weight, bias, params.padding, params.stride, params.groups);
+        input.device().type(), input, weight, bias, params.stride, params.padding, params.groups);
```
Comment (dzhulgakov) [inline aten/src/ATen/native/Convolution.cpp]: wait, was it a bug? do we have a unittest for it?

Comment (AshkanAliabadi) [reply aten/src/ATen/native/Convolution.cpp]: This is a crazy one.  Here's my explanation of how I think it was working: https://github.com/pytorch/pytorch/pull/27117#discussion_r331209266

File: aten/src/ATen/CMakeLists.txt [hunk 2]
```diff
@@ -211,7 +211,7 @@ endif()
 
 if(AT_NNPACK_ENABLED)
   include_directories(${NNPACK_INCLUDE_DIRS})
-  list(APPEND ATen_CPU_DEPENDENCY_LIBS nnpack) # cpuinfo is added below
+  list(APPEND ATen_CPU_DEPENDENCY_LIBS nnpack nnpack_reference_layers) # cpuinfo is added below
```
Comment (AshkanAliabadi) [inline aten/src/ATen/CMakeLists.txt]: NNPACK does not support strided convolutions on batches (used during training,) hence we fallback to the slower reference implementation here.

File: aten/src/ATen/native/NNPACK.cpp [hunk 3]
```diff
@@ -243,16 +252,35 @@ Tensor _nnpack_spatial_convolution(
...
+          output_channels,
+          input_size,
+          input_padding,
+          kernel_size,
+          output_subsample,
+          input_.data_ptr<float>(),
+          weight.data_ptr<float>(),
+          bias_.data_ptr<float>(),
+          output.data_ptr<float>(),
+          nnpack_threadpool());
```
Comment (AshkanAliabadi) [inline aten/src/ATen/native/NNPACK.cpp]: NNPACK does not support strided convolutions on batches (used during training,) hence we fallback to the slower reference implementation here.

**Review by AshkanAliabadi — COMMENTED** [review]
Pending CI, it seems that I finally managed to get this working.  This is really not an ideal solution in my opinion and definitely not the direction we should be taking PyTorch mobile in, in the long run.  I think the ideal solution would be to have one single mobile-focused backend, just like the reference or vendor-provided backends we currently have, that handles floating point and quantized operations both for NHWC and NCHW layout but efficiently for mobile.  That is a considerable undertaking though, and if we want to close the gap in the meanwhile, we need short-term solutions like this unfortunately.

**Comment by AshkanAliabadi** [comment]
Dependent on https://github.com/Maratyszcza/NNPACK/pull/172.

**Pull request closed without merging**